crates/fluvio/src/producer/mod.rs

```diff
@@ -33,12 +45,224 @@ use crate::producer::partitioning::{Partitioner, SiphashRoundRobinPartitioner, P
 /// you can send events to the topic, choosing which partition
 /// each event should be delivered to.
 pub struct TopicProducer {
-    topic: String,
-    pool: Arc<SpuPool>,
-    partitioner: Box<dyn Partitioner + Send + Sync>,
+    inner: Arc<InnerTopicProducer>,
     #[cfg(feature = "smartengine")]
     pub(crate) smartmodule_instance: Option<Arc<RwLock<Box<dyn SmartModuleInstance>>>>,
 }
+
+/// Struct that is responsible for sending produce requests to the SPU in a given partition.
+struct PartitionProducer {
+    topic: String,
+    partition: PartitionId,
+    spu_pool: Arc<SpuPool>,
+    batches_lock: Arc<Mutex<VecDeque<ProducerBatch>>>,
+    linger: Duration,
+}
+
+impl PartitionProducer {
+    fn new(
+        topic: String,
+        partition: PartitionId,
+        spu_pool: Arc<SpuPool>,
+        batches_lock: Arc<Mutex<VecDeque<ProducerBatch>>>,
+        linger: Duration,
+    ) -> Self {
+        Self {
+            topic,
+            partition,
+            spu_pool,
+            batches_lock,
+            linger,
+        }
+    }
+
+    fn shared(
+        topic: String,
+        partition: PartitionId,
+        spu_pool: Arc<SpuPool>,
+        batches: Arc<Mutex<VecDeque<ProducerBatch>>>,
+        linger: Duration,
+    ) -> Arc<Self> {
+        Arc::new(PartitionProducer::new(
+            topic, partition, spu_pool, batches, linger,
+        ))
+    }
+
+    /// Flush all the batches that are full or have reached the linger time.
+    /// If force is set to true, flush all batches regardless of linger time.
+    async fn flush(&self, force: bool) -> Result<()> {
+        let replica_key = ReplicaKey::new(self.topic.clone(), self.partition);
+        let partition_spec = self
+            .spu_pool
+            .metadata
+            .partitions()
+            .lookup_by_key(&replica_key)
+            .await?
+            .ok_or_else(|| FluvioError::PartitionNotFound(self.topic.to_string(), self.partition))?
+            .spec;
+        let leader = partition_spec.leader;
+
+        let spu_socket = self
+            .spu_pool
+            .create_serial_socket_from_leader(leader)
+            .await?;
+
+        let mut batches_ready = vec![];
```

Comment:
shouldn't batch be already calculated at this time?

Reply:
There could be batches in the VecDeq that are not ready to be flushed (they are not full and their creation time is smaller than the linger time